File: binary_search/java/TestBinarySearch.java
import java.util.Random;
import java.util.Arrays;

public class TestBinarySearch
{
	public static final int size = 1000;
	public static Random myrand = new Random();
	public static int[] setUp(int size)
	{
		int[] array = new int[size];
		int i, target;
		for(i = 0; i < size; i++)
			array[i] = myrand.nextInt();
		return array;
	}
	public static void testBinarySearch()
	{
		int[] array = new int[size];
		for(int i = 0; i < size; i++)
			array[i] = i;
		for(int i = 0; i < 100; i++)
		{
			int target = myrand.nextInt() % (size * 2);
			if(target >= size)
				assert BinarySearch.search(array, target) == -1;
			else
				assert BinarySearch.search(array, target) == target;
		}
	}
	public static void testLower()
	{
		int[] array = setUp(size);
		Arrays.sort(array);
		int lower = 0;
		for(int i = 0; i < 100; i++)
		{
			int target = myrand.nextInt();
			for(int j = 0; j < size; j++)
				if(array[j] >= target)
				{
					lower = j - 1;
					break;
				}
			assert BinarySearch.lower(array, target) == lower;
		}
	}
	public static void testHigher()
	{
		int[] array = setUp(size);
		Arrays.sort(array);
		int upper = 0;
		for(int i = 0; i < 100; i++)
		{
			int target = myrand.nextInt();
			for(int j = size - 1; j >= 0; j--)
				if(array[j] <= target)
				{
					upper = j + 1;
					break;
				}
			assert BinarySearch.higher(array, target) == upper;
		}
	}
	public static void main(String[] args)
	{
	}
}

File: binary_search/py/TestBinarySearch.py
from binary_search import binary_search, lower, higher, equal_range
import bisect
import unittest
import random

class TestBinarySearch(unittest.TestCase):
    def setUp(self):
        self.times = 100
        self.size = 30000
        self.array = [random.randrange(self.size) for i in range(self.size)]
        self.array.sort()

    def test_equal_range(self):
        for i in range(self.times):
            target = random.randrange(int(self.size * 1.10))
            retval = equal_range(self.array, target) 
            left = bisect.bisect_left(self.array, target)
            if 0 <= left < len(self.array) and self.array[left] == target:
                right = bisect.bisect(self.array, target) - 1
            else:
                left = right = -1
            self.assertEqual(retval, (left, right))

    def test_higher(self):
        for i in range(self.times):
            target = random.randrange(int(self.size * 1.10))
            retval = higher(self.array, target) 
            expect = bisect.bisect(self.array, target)
            self.assertEqual(retval, expect)

    def test_lower(self):
        for i in range(self.times):
            target = random.randrange(int(self.size * 1.10))
            retval = lower(self.array, target) 
            expect = bisect.bisect_left(self.array, target) - 1
            self.assertEqual(retval, expect)

    def test_binary_search(self):
        self.array = [i for i in range(self.size)]
        for i in range(self.times):
            target = random.randrange(int(self.size * 1.10))
            retval = binary_search(self.array, target) 
            if 0 <= retval < len(self.array):
                self.assertEqual(self.array[retval], target)
            else:
                self.assertEqual(retval, -1)
                self.assertTrue(target < self.array[0] or self.array[-1] < target)

if __name__ == '__main__':
    unittest.main()
